lightning/src/chain/package.rs

```diff
@@ -369,25 +430,107 @@ pub(crate) struct HolderHTLCOutput {
 	/// Defaults to 0 for HTLC-Success transactions, which have no expiry
 	cltv_expiry: u32,
 	channel_type_features: ChannelTypeFeatures,
+	htlc_descriptor: Option<HTLCDescriptor>,
 }
 
 impl HolderHTLCOutput {
-	pub(crate) fn build_offered(amount_msat: u64, cltv_expiry: u32, channel_type_features: ChannelTypeFeatures) -> Self {
-		HolderHTLCOutput {
-			preimage: None,
+	pub(crate) fn build(htlc_descriptor: HTLCDescriptor) -> Self {
+		let amount_msat = htlc_descriptor.htlc.amount_msat;
+		let channel_type_features = htlc_descriptor.channel_derivation_parameters
+			.transaction_parameters.channel_type_features.clone();
+		Self {
+			preimage: if htlc_descriptor.htlc.offered {
+				None
+			} else {
+				Some(htlc_descriptor.preimage.expect(""))
+			},
 			amount_msat,
-			cltv_expiry,
+			cltv_expiry: if htlc_descriptor.htlc.offered {
+				htlc_descriptor.htlc.cltv_expiry
+			} else {
+				0
+			},
 			channel_type_features,
+			htlc_descriptor: Some(htlc_descriptor),
 		}
 	}
 
-	pub(crate) fn build_accepted(preimage: PaymentPreimage, amount_msat: u64, channel_type_features: ChannelTypeFeatures) -> Self {
-		HolderHTLCOutput {
-			preimage: Some(preimage),
-			amount_msat,
-			cltv_expiry: 0,
-			channel_type_features,
+	pub(crate) fn get_htlc_descriptor<ChannelSigner: EcdsaChannelSigner>(
+		&self, onchain_tx_handler: &mut OnchainTxHandler<ChannelSigner>, outp: &::bitcoin::OutPoint,
```

Comment:
Looks like `OnchainTxHandler` doesn't need to be a mutable reference anywhere that it is passed in this file.